docs/overview.rst

```diff
@@ -1,11 +1,410 @@
+   :class:`collections.OrderedDict` instances if the source order is important
+   to you.
+
+Datasets themselves are stateless objects (as opposed to, say, an open file
+handle, or an iterator object). In order to request data from the dataset, we
+need to ask it to instantiate some stateful object with which it will interact.
+This is done through the :meth:`Dataset.open` method:
+
+>>> state = dataset.open()
+>>> print(state.__class__.__name__)
+imap
+
+We see that in :class:`IterableDataset`'s case the state is an iterator object.
+We can now visit the examples this dataset contains using its
+:meth:`get_data` method.
+
+>>> print(dataset.get_data(state=state))
+(array([[ 47, 211],
+       [ 38,  53]]), array([0]))
+>>> while True:
+...     try:
+...         __ = dataset.get_data(state=state)
+...     except StopIteration:
+...         print('Iteration over')
+...         break
+Iteration over
+
+Eventually, the iterator is depleted and it raises a :class:`StopIteration`
+exception. We can iterate over the dataset again by requesting a fresh iterator
+through the dataset's :meth:`reset` method.
+
+>>> state = dataset.reset(state=state)
+>>> print(dataset.get_data(state=state))
+(array([[ 47, 211],
+       [ 38,  53]]), array([0]))
+
+When you're done, don't forget to call the dataset's :meth:`close` method on
+the state. This has the effect of cleanly closing the state (e.g. if the state
+is an open file handle, :meth:`close` will close it).
+
+>>> dataset.close(state=state)
+
+IndexableDataset
+^^^^^^^^^^^^^^^^
+
+The :class:`IterableDataset` implementation is pretty minimal. For instance, it
+only lets you iterate sequentially and examplewise over your data.
+
+If your data happens to be indexable (e.g. a :class:`list`, or a
+:class:`numpy.ndarray`), then :class:`IndexableDataset` will let you do much
+more.
+
+We instantiate :class:`IndexableDataset` just like :class:`IterableDataset`.
+
+>>> from fuel.datasets import IndexableDataset
+>>> dataset = IndexableDataset(
+...     indexables=OrderedDict([('features', features), ('targets', targets)]),
+...     axis_labels=OrderedDict([('features', ('batch', 'height', 'width')),
+...                              ('targets', ('batch', 'index'))]))
+
+The main advantage of :class:`IndexableDataset` over :class:`IterableDataset`
+is that it allows random access of the data it contains. In order to do so, we
+need to pass an additional ``request`` argument to :meth:`get_data` in the form
+of a list of indices.
+
+>>> state = dataset.open()
+>>> print('State is {}.'.format(state))
+State is None.
+>>> print(dataset.get_data(state=state, request=[0, 1]))
+(array([[[ 47, 211],
+        [ 38,  53]],
+<BLANKLINE>
+       [[204, 116],
+        [152, 249]]]), array([[0],
+       [3]]))
+>>> dataset.close(state=state)
+
+See how :class:`IndexableDataset` returns a ``None`` state: this is because
+there's no actual state to maintain in this case.
+
+Restricting sources
+^^^^^^^^^^^^^^^^^^^
```

Comment:
I agree it is not simple to find a fully satisfying solution here.

I suggest to introduce the generic dataset and all its methods in the "generic section", specifying that, given it is an abstract class, all the code will be shown using one the classes that implement it, but that the same code would work for any other class that implement the Dataset class. 
This would have these advantages:
* to show working code 
* to have in the generic section what applies to all the datasets
* to keep the other sections (e.g. `IterableDataset`) "straight to the point", showing only what is really specific to them